iceoryx_utils/include/iceoryx_utils/internal/lockfree_queue/index_queue.hpp

```diff
@@ -0,0 +1,149 @@
+
+    using value_t = ValueType;
+
+    struct ConstructFull_t
+    {
+    };
+
+    struct ConstructEmpty_t
+    {
+    };
+
+    // @todo: a compile time check whether Index is actually lock free would be nice
+    // note: there is a way  with is_always_lock_free in c++17 (which we cannot use here)
+
+    static constexpr ConstructFull_t ConstructFull{};
+    static constexpr ConstructEmpty_t ConstructEmpty{};
+
+    IndexQueue(const IndexQueue&) = delete;
+    IndexQueue(IndexQueue&&) = delete;
+    IndexQueue& operator=(const IndexQueue&) = delete;
+    IndexQueue& operator=(IndexQueue&&) = delete;
+
+    /// @brief constructs an empty IndexQueue
+    IndexQueue(ConstructEmpty_t = ConstructEmpty);
+
+    /// @brief constructs IndexQueue filled with all indices 0,1,...capacity-1
+    IndexQueue(ConstructFull_t);
+
+    /// @brief get the capacity of the IndexQueue
+    /// @return capacity of the IndexQueue
+    /// threadsafe, lockfree
+    constexpr uint64_t capacity();
+
+    /// @brief check whether the queue is empty
+    /// @return true iff the queue is empty
+    /// note that if the queue is used concurrently it might
+    /// not be empty anymore after the call
+    /// (but it was at some point during the call)
+    bool empty();
+
+    // The advantage of the UniqueIndex interface is that it prevents us from returning
+    // an index multiple times by design, i.e. it enforces that only indices popped from
+    // an IndexQueue can be returned.
+    // This works by preventing copies and construction of UniqueIndex outside of the IndexQueue.
+    // In particular, the user is free to get and copy the raw index, but he *cannot* construct a new
+    // UniqueIndex from it.
+
+    /// @brief push index into the queue in FIFO order
+    /// always succeeds if the UniqueIndex to be pushed is popped from another equallysized
+    /// IndexQueue (and should only be used this way)
+    /// threadsafe, lockfree
+    /// @param index to be pushed, any index can only be pushed once by design
+    void push(UniqueIndex& index);
+
+    /// @brief tries to remove index in FIFO order
+    /// threadsafe, lockfree
+    /// @return valid UniqueIndex if removal was successful (i.e. queue was not empty),
+    /// invalid UnqiueIndex otherwise
+    UniqueIndex pop();
+
+    /// @brief tries to remove index in FIFO order iff the queue is full
+    /// threadsafe, lockfree
+    /// @return valid UniqueIndex if removal was successful (i.e. queue was full),
+    /// invalid UnqiueIndex otherwise
+    UniqueIndex popIfFull();
+
+  private:
+    using Index = CyclicIndex<Capacity>;
+
+    using Cell = std::atomic<Index>;
+    Cell m_cells[Capacity];
+
+    std::atomic<Index> m_readPosition;
+    std::atomic<Index> m_writePosition;
+
+  private:
+    Index loadNextReadPosition() const;
+    Index loadNextWritePosition() const;
+    Index loadValueAt(Index position) const;
+    bool tryToPublishAt(Index writePosition, Index& oldValue, Index newValue);

```

Comment:
Please add const to the arguments

Reply:
I am against doing this for value types all over the place without critical analysis. We will refine our static analysis ruleset to drop this requirement should this be a problem. This is not really needed and as far as I know does only very rarely offer optimization opportunities to the compiler it would not use anyway. @mossmaurice

Reply:
I agree with you. You can hardly ever generalise. The philosophical discussion about clean code is very important as well. However, before running off the track: Please just ran QAC++ tomorrow and see if coding rules are violated here. Const correctness is not always the highest good IMHO.

Nitpick: Before claiming that abstraction costs performance, have you measured it? ;-)

Reply:
Abstraction is almost never free, sometimes it does not cost you performance but you still pay something, such as longer compile time. But here it is mainly about const ness of value types, which has no real upsides and is just cluttering the interface. Still, there is no universal agreement. See for example this discussion https://stackoverflow.com/questions/8532095/is-it-a-good-idea-of-maintaining-const-ness-as-much-as-possible. Just to be clear, I am all for const correctness of methods and reference arguments (but still think C++ is to verbose here). But most importantly it is an information for the caller and, for references, helps to enforce the contract. For value types it provides not many optimization opportunities for the compiler, since it has to do an analysis if the value is really not changed in any case (constness can be circumvented).